Detectors/PHOS/calib/PHOSCalibWorkflow/src/PHOSBadMapCalibDevice.cxx

```diff
@@ -0,0 +1,379 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "PHOSCalibWorkflow/PHOSCalibCollector.h"
+#include "DataFormatsPHOS/TriggerRecord.h"
+#include "Framework/ConfigParamRegistry.h"
+#include "DetectorsCalibration/Utils.h"
+#include "Framework/ControlService.h"
+
+#include "FairLogger.h"
+#include <fstream> // std::ifstream
+
+using namespace o2::phos;
+
+void PHOSCalibCollector::init(o2::framework::InitContext& ic)
+{
+
+  mEvent = 0;
+  //Create output histograms
+  const int nChannels = 14336; //4 full modules
+  const int nMass = 150.;
+  float massMax = 0.3;
+  //First create all histograms then get direct pointers
+  mHistos.emplace_back("hReInvMassPerCell", "Real inv. mass per cell", nChannels, 0., nChannels, nMass, 0., massMax);
+  mHistos.emplace_back("hMiInvMassPerCell", "Mixed inv. mass per cell", nChannels, 0., nChannels, nMass, 0., massMax);
+
+  int npt = 45;
+  float xpt[46] = {0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1., 1.2, 1.4, 1.6, 1.8, 2., 2.2, 2.4, 2.6, 2.8, 3., 3.4, 3.8, 4.2, 4.6, 5.,
+                   5.5, 6., 6.5, 7., 7.5, 8., 9., 10., 12., 14., 16., 18., 20., 24., 28., 32., 36., 40., 50., 55., 60.};
+  float massbins[nMass + 1];
+  for (int i = 0; i <= nMass; i++) {
+    massbins[i] = i * massMax / nMass;
+  }
+  mHistos.emplace_back("hReInvMassNonlin", "Real inv. mass vs Eclu", nMass, massbins, npt, xpt);
+  mHistos.emplace_back("hMiInvMassNonlin", "Mixed inv. mass vs Eclu", nMass, massbins, npt, xpt);
```

Comment:
note that filling histos with custom binning axis is much slower (not to speak that root histos are much slower to fill than e.g. boost ones). Is it really needed?